### ISSUE TITLE
[Feature request] A suitable test problem for Adam

### ISSUE THREAD

**Issue by AjayTalati** [torch/optim#47]
Could I request a test problem for the Adam optimizer, just to understand how it works better? Thanks :+1: 

I tried to use the new adam optimizer on the rosenbrock test problem in `optim`, used for adagrad, but it does'nt seem to work? I can't get it to work with a range of different `config` parameters for rosenberg, or for the ML problem I'm working on - simple copy tasks using the neural Turing machine?  

I expect I've misunderstood the [adam paper](http://arxiv.org/pdf/1412.6980.pdf), so I'm doing something wrong/really dumb -- _does the objective necessarily have to be stochastic_ -- for adam to be applied? 

If so then I expect it would'nt work for rosenbrock, and the LSTM used in the neural Turing machine which is which is deterministic? 

My failed rosenbrock attempt

require 'torch'
require 'optim'
require 'rosenbrock'
require 'l2'
x = torch.Tensor(2):fill(0)
fx = {}
config_adagrad = {learningRate=1e-1}

config_adam = adam_config = {
    learningRate = 1e-6,
    beta1 = 0.01,
    beta2 = 0.001
}

for i = 1,10001 do
      --x,f=optim.adagrad(rosenbrock,x,config_adagrad)
      x,f=optim.adam(rosenbrock,x,config_adam)
      if (i-1)%1000 == 0 then
           table.insert(fx,f[1])
      end
end

print()
print('Rosenbrock test')
print()
print('x=');print(x)
print('fx=')

OUTPUT

Rosenbrock test	
x=	
0.01 *
 2.0243
 0.0523
[torch.DoubleTensor of dimension 2]

fx=	
1	1	
1001	0.96578919291079	
2001	0.96476690379526	
3001	0.96406793828219	
4001	0.96344671327807	
5001	0.96285032358036	
6001	0.96226242754115	
7001	0.96167791102736	
8001	0.96111328576091	
9001	0.96050921334115	
10001	0.9599246681675

**Comment by AjayTalati** [comment]
Update - I've only read about Adam being used for models with latent variables, in particular for the class of variational autoencoders. 

Here are the papers with mention/imply it's use,



So if there's any torch implementations of training loops for these - please post them here?

**Comment by soumith** [comment]
let's keep the issue open, it's a fair request.

**Comment by AjayTalati** [comment]
OK. I'm trying to get it to work for the [train-a-digit-classifier](https://github.com/torch/demos/blob/master/train-a-digit-classifier/train-on-mnist.lua), in the demos repo. 

This is the simplest example they give in their paper -- they don't give any numbers, but [sec 6.1](http://arxiv.org/pdf/1412.6980.pdf) gives a figure for logistic regression, which I'm trying to reproduce? 

It takes about 10 mins for one iteration, and I guess they use the cost `ClassNLLCriterion()` ??? I'm still confused :(

**Comment by AjayTalati** [comment]
update - I just saw the author of `adam.lua` louissmit has a package [VBNN](https://github.com/louissmit/VBNN) which tries to do the above so I'll try using that?

**Comment by AjayTalati** [comment]
MANY THANKS TO IVO DANIHELKA for checking the code and finding the square root error :+1:  :+1:  :+1:   

I though I was going insane for a while?

The Rosenbrock test with the default  `adam` parameters

Rosenbrock test	
x=	
 1.0000
 1.0000
[torch.DoubleTensor of dimension 2]

fx=	
1	1	
1001	0.65439962865955	
2001	0.40238118820709	
3001	0.22511224420514	
4001	0.10887403144721	
5001	0.041717668856584	
6001	0.010800627503131	
7001	0.0014015139144866	
8001	5.3520254822433e-05	
9001	2.5441127189386e-07	
10001	3.1952390238754e-09

**Comment by AjayTalati** [comment]
OK I ran adam on a torch implementation of the NTM overnight and it gives reasonable convergence, but 50K is too long compared to RMSprop while learns a copy task after 10K iterations.

I saw Shawn Tan's feedfroward python implementation uses [ciriculum learning](https://blog.wtf.sg/2014/11/11/neural-turing-machines-copy-task/), which seems like a bright idea :+1:  

It's a total hack !!!! But training the neural turing machine, for very short sequences <5, and embeeding it in Whetlab might give one method of getting a sensible learning rate , and lambda and epsilon for using adam with the ntm.

Any advice very gratefully recieved!

**Comment by AjayTalati** [comment]
There's some discussion of a good lambda setting here - https://twitter.com/ogrisel/status/568316753593434112

Apparently there might be a type mistake in v2 of the paper as stated here

https://twitter.com/ogrisel/status/568316753593434112

**Comment by soumith** [comment]
This PR: https://github.com/torch/optim/pull/46 addressed that issue according to those twitter comments already.

**Comment by AjayTalati** [comment]
Thanks, Sousmith! 

May I ask your advice -- is whetlab the way to go for general hyperparameter tuning? There's lua MNIST [code](https://github.com/whetlab/Whetlab-Lua-Client/blob/master/examples/tutorial_example.lua) that comes with it, (it's a naff MLP model), and they use SGD.  

I'm happy to do all the 'donkey work' to get a good test problem for adam working in torch -- but so far I have'nt got anything _convincing_ going for the ntm problem I'm working on, (or even the original adam paper)?  

I'm not convinced yet of any Adam vs RMSprop benchmark test -- arvix papers are really great, but they're much more likely to get published if they provide clean open source (torch) code to back them up? Same goes for the neural Turing machine?

**Comment by soumith** [comment]
How about you pick mnist, cifar and imagenet in that order. Compare sgd, rmsprop, adam, lbfgs etc.

**Comment by AjayTalati** [comment]
Are there any hyperparameter tunning packages available in torch, or open-sourced that I could try to port over? Perhaps in R

Does'nt have to be state of the art, just simple and stable?

Whetlab has a 5000 hyper-parameter suggestion per month limit for their free public beta.

**Comment by soumith** [comment]
there's whetlab for torch, but you can use spearmint as well

**Comment by AjayTalati** [comment]
Thanks Soumith!

The dp package here in torch, has some [hyperparameter optimization](http://dp.readthedocs.org/en/latest/neuralnetworktutorial/index.html#hyperoptimizing). Have you seen it, or even used it for this?

I'm trying to read the [code](https://github.com/nicholas-leonard/dp/tree/master/hyper), to see how it works - looks like it could be used to implement Gaussian process Bayesian optimization using Gaussian processes? 

Package looks like it could be useful for validating new code?

It's not obvious, (to me anyway), how to switch optimizers to adam? I'll ask Nicholas Leonard

**Comment by nicholas-leonard** [comment]
The dp code for hyperparameter tunning just samples from a user-defined prior. No gaussian process. Personally, I just hand tune hyper-parameters as they otherwise take too much time and its more difficult to learn what is actually going on. Of course, I would love to have something like whetlab to hyper-optimize the code.

**Comment by AjayTalati** [comment]
The [Adam paper](http://arxiv.org/pdf/1412.6980.pdf) was updated on Tuesday 3rd March. 

Thanks Durk & Jimmy :+1: :+1: :+1: 

So I'm checking the code here, but I'm not the cleverest person, (bit thick to be honest), maybe someone who's smart can also check, `adam.lua`?

**Comment by AjayTalati** [comment]
So with the default parameters of the latest version of the paper this is what I think the code should would be?

-- Decay the first moment running average coefficient
`local bt1 = beta1*lambda^(state.t - 1)`

-- Update biased first moment estimates --state.m:mul(bt1):add(1-bt1, dfdx)
`state.m = torch.add( torch.mul(state.m, bt1) , torch.mul(dfdx, 1-bt1) )` 

-- Update biased second raw moment estimate --state.v:mul(beta2):addcmul(1-beta2, dfdx, dfdx)
`state.v = torch.add( torch.mul(state.v, beta2) , torch.mul( torch.cmul(dfdx, dfdx) , 1-beta2 ) )` 

state.denom:copy(state.v):sqrt():add(epsilon)

`local biasCorrection1 = 1 - beta1^state.t`
`local biasCorrection2 = 1 - beta2^state.t`
`local stepSize = lr * math.sqrt(biasCorrection2)/biasCorrection1`

--update x
`x:add( torch.mul( torch.cdiv( state.m, state.denom ) , -stepSize ) )`

But it still does'nt work for me?

**Comment by AjayTalati** [comment]
SUCCESS !!!!!! 

I finally got it to work - I think?

I believe there's a type error in the paper - but it's not a big deal. In the 3rd March version version 4 paper, I think the sign in the update rule should be a `plus` not a `minus`. 

That is, on page 2, on the fifth line from the bottom should be,

theta_t <- theta_t-1 + alpha_t . m_t / (sqrt(v_t) + epsilon )

not 

theta_t <- theta_t-1 - alpha_t . m_t / (sqrt(v_t) + epsilon )

Anyhow - I got it to work for my problem so I'm very very happy :+1:
